Kernel/ACPI/DMIDecoder.h

```diff
@@ -0,0 +1,1342 @@
+    u8 protocol_records_count;
+    ProtocolRecordData protocol_records[];
+} __attribute__((packed));
+
+enum class ManagementControllerHostInterfaceProtocolType {
+    IPMI = 0x2,
+    MCTP = 0x3,
+    RedfishOverIP = 0x4
+};
+
+struct TPMDevice { // Type 43
+    TableHeader h;
+    char vendor_id[4];
+    u8 major_spec_version;
+    u8 minor_spec_version;
+    u32 firmware_version_1;
+    u32 firmware_version_2;
+    u8 description_str_number;
+    u64 characteristics;
+    u32 oem_defined;
+} __attribute__((packed));
+
+enum class TPMDeviceCharacteristics {
+    Characteristics_not_supported = (1 << 2),
+    Family_Configurable_1 = (1 << 3), // Family configurable via firmware update; for example, switching between TPM 1.2 and TPM 2.0.
+    Family_Configurable_2 = (1 << 4), // Family configurable via platform software support, such as BIOS Setup; for example, switching between TPM 1.2 and TPM 2.0.
+    Family_Configurable_3 = (1 << 5), // Family configurable via OEM proprietary mechanism; for example, switching between TPM 1.2 and TPM 2.0.
+};
+
+struct ProcessorSpecificBlock {
+    u8 block_length;
+    u8 processor_type;
+    u8 processor_specific_data[];
+} __attribute__((packed));
+
+struct ProcessorAdditionalInfo { // Type 44
+    TableHeader h;
+    u16 referenced_handle;
+    ProcessorSpecificBlock blocks[];
+} __attribute__((packed));
+
+enum class ProcessorArchitectureType {
+    IA32 = 0x1,
+    x86_64 = 0x2,
+    Itanium = 0x3,
+    ARM32bit = 0x4,
+    ARM64bit = 0x5,
+    RISC_V_32bit = 0x6,
+    RISC_V_64bit = 0x7,
+    RISC_V_128bit = 0x8
+};
+
+struct Inactive { // Type 126
+    TableHeader h;
+} __attribute__((packed));
+
+struct EndOfTable { // Type 127
+    TableHeader h;
+} __attribute__((packed));
+}
+
+class DMIDecoder {
+public:
+    static DMIDecoder& the();
+    static void initialize();
+    static void initialize_untrusted();
+    Vector<SMBIOS::PhysicalMemoryArray*>& get_physical_memory_areas();
+    bool is_reliable();
+    u64 get_bios_characteristics();
+
+private:
+    void enumerate_smbios_tables();
+    SMBIOS::TableHeader* get_next_physical_table(SMBIOS::TableHeader& p_table);
+    SMBIOS::TableHeader* get_smbios_physical_table_by_handle(u16 handle);
+    SMBIOS::TableHeader* get_smbios_physical_table_by_type(u8 table_type);
+    char* get_smbios_string(SMBIOS::TableHeader& p_table, u8 string_number);
+    size_t get_table_size(SMBIOS::TableHeader& table);
+
+    explicit DMIDecoder(bool trusted);
+    void initialize_parser();
+
+    OwnPtr<SMBIOS::EntryPoint32bit> find_entry32bit_point();
+    OwnPtr<SMBIOS::EntryPoint64bit> find_entry64bit_point();
+
+    void mmap(VirtualAddress vaddr, PhysicalAddress paddr, u32 length);
+    void mmap_region(Region& region, PhysicalAddress paddr);
+
+    OwnPtr<SMBIOS::EntryPoint32bit> m_entry32bit_point;
+    OwnPtr<SMBIOS::EntryPoint64bit> m_entry64bit_point;
+    OwnPtr<SMBIOS::TableHeader> m_structure_table;
```

Comment:
Are we sure we want `OwnPtr` semantics for these?
`OwnPtr` means that we will automatically call `delete` on them when they go out of scope.
If these are just pointers into persistent data that someone else keeps alive, they should just be raw `T*` instead.

Reply:
What do you mean by 'going out of the scope' here? anyway, my initial thought was to actually use those, but since there is someone that owns those pointers...  I changed it.

Reply:
In this case, since they are members of `DMIDecoder`, it means we will call `~OwnPtr()` on them automatically from `~DMIDecoder()` when it's destroyed.
Even if the `DMIDecoder` is never destroyed, we shouldn't use `OwnPtr` since the main purpose of `OwnPtr` is to ensure that `delete` gets called :)

Reply:
OK then, I see why it's better this way :)
I'll fix it.
Also, I guess this is the same fate for `OwnPtr`s being used in the ACPIStaticParser class.